uc-crux-llvm/src/UCCrux/LLVM/Errors/Unimplemented.hs

```diff
@@ -0,0 +1,89 @@
+{-
+Module       : UCCrux.LLVM.Errors.Unimplemented
+Description  : Dealing with unimplemented features
+Copyright    : (c) Galois, Inc 2021
+License      : BSD3
+Maintainer   : Langston Barrett <langston@galois.com>
+Stability    : provisional
+-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE Trustworthy #-}
+
+module UCCrux.LLVM.Errors.Unimplemented
+  ( Unimplemented (..),
+    ppUnimplemented,
+    unimplemented,
+    catchUnimplemented,
+    assertUnimplemented,
+  )
+where
+
+{- ORMOLU_DISABLE -}
+import Control.Exception (SomeException, try, displayException)
+import Panic hiding (panic)
+import qualified Panic
+{- ORMOLU_ENABLE -}
+
+data Unimplemented
+  = VarArgsFunction
+  | VarArgsFunctionType
+  | VoidType
+  | OpaqueType
+  | UnsupportedType
+  | MetadataType
+  | GeneratingArrays
+  | IndexCursor
+  | ConstrainGlobal
+  deriving (Bounded, Enum, Eq, Ord)
+
+ppUnimplemented :: Unimplemented -> String
+ppUnimplemented =
+  \case
+    VarArgsFunction -> "Exploring variable-arity functions"
+    VarArgsFunctionType -> "Variable-arity function (pointer) types in globals or arguments"
+    VoidType -> "Void types in globals or arguments"
+    OpaqueType -> "Opaque (undefined) types in globals or arguments"
+    UnsupportedType -> "Unsupported types in globals or arguments"
+    MetadataType -> "LLVM metadata types in globals or arguments"
+    GeneratingArrays -> "Arrays in globals or arguments"
+    IndexCursor -> "Deduced preconditions on array elements"
+    ConstrainGlobal -> "Constraints on a global variable"
+
+instance PanicComponent Unimplemented where
+  panicComponentName _ = "uc-crux-llvm"
+  panicComponentIssues _ = "https://github.com/GaloisInc/crucible/issues"
+
+  {-# NOINLINE panicComponentRevision #-}
+  panicComponentRevision = $useGitRevision
+
+unimplemented ::
+  HasCallStack =>
+  -- | Short name of where the error occured
+  String ->
+  -- | What's the not-yet supported thing?
+  Unimplemented ->
+  a
+unimplemented where_ what =
+  Panic.panic
+    what
+    where_
+    [ ppUnimplemented what,
+      "is not yet implemented as a feature of crux-llvm bugfinding mode.",
+      "If this feature would be useful to you, please report this error."
+    ]
+
+catchUnimplemented :: IO a -> IO (Either (Panic.Panic Unimplemented) a)
```

Comment:
I don't love this pattern where we catch `panic`; the idea behind panic is that it should be an unrecoverable error that results from a programming error that we believe should not happen.  This use is pretty distinct from that.

Let's find another way to do this without piggy-backing on `panic`.  If we want to add a new feature to the panic package for this kind of use, I'd certainly be open to that idea.

Reply:
Yeah, I was just being a bit lazy here. I'll make a new exception type with similar fields.

Reply:
Ah, when looking at this code again I remembered what I liked about `Panic`: 

- it links to the issue tracker and asks the user to report the issue, making it clear that it *is* a problem with the tool, not a user error
- It includes the git revision
- It captures the callstack and prints it in the report

These are nice features and would take a bit of effort to re-engineer, so I'll make a follow-up issue for this.

Reply:
https://github.com/GaloisInc/crucible/issues/695